### PR TITLE
Add FlaxCLIPTextModelWithProjection

### PR DESCRIPTION
**Pull request by pcuenca**
# What does this PR do?

`FlaxCLIPTextModelWithProjection` is necessary to support the Flax port of Stable Diffusion XL: https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/fb6d705fb518524cabc79c77f13a0e7921bcab3a/text_encoder_2/config.json#L3

I can add some tests, if necessary, after this approach is validated.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @patil-suraj @sanchit-gandhi @younesbelkada

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Should we maybe for now just add it in a subfolder of sdxl in diffusers here: https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/stable_diffusion_xl  instead of having to rely on `transformers` here? I'm not 100% convinced this model is really needed for core transformers usage.

Would also not force the user to have to install transformers from main :-)

**Comment by pcuenca** [comment]
> Should we maybe for now just add it in a subfolder of sdxl in diffusers here: https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/stable_diffusion_xl instead of having to rely on `transformers` here? I'm not 100% convinced this model is really needed for core transformers usage.

The [PyTorch version of the same model was added 9 months ago](https://github.com/huggingface/transformers/blob/bd90cda9a6bb4723515c17df1192e53abc8e36e3/src/transformers/models/clip/modeling_clip.py#L1198), so I assumed it was ok.

But sure, we can do that. In that case, how do we deal with it?
- Change the library to `diffusers` here: https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/main/model_index.json#L15. Unless I'm mistaken, then we'd need to distribute the flax weights separately, or use a branch.
- Create a hack in diffusers to map the library.

> 
> Would also not force the user to have to install transformers from main :-)

Yes, of course, this was meant as the long-term solution.

**Comment by patrickvonplaten** [comment]
Ah yeah good point JAX & PyTorch share the same config - this will become complicated indeed then. Ok let's try to get it merged here. CLIP is important enough to be merged to `transformers` indeed

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 1]
```diff
@@ -1007,6 +1007,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
...
+        )
+
+        pooled_output = text_outputs.pooler_output
+        text_embeds = self.text_projection(pooled_output)
+
+        if not return_dict:
+            return (text_embeds, pooled_output) + text_outputs[2:]
+
+        return FlaxBaseModelOutputWithPooling(
+            last_hidden_state=text_embeds,
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply src/transformers/models/clip/modeling_flax_clip.py:1045]: I initially hesitated doing the same, but it makes sense, will fix :)

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 2]
```diff
@@ -155,6 +155,36 @@
 """
 
 
+@flax.struct.dataclass
+class FlaxCLIPTextModelOutput(ModelOutput):
+    """
+    Base class for text model's outputs that also contains a pooling of the last hidden states.
+
+    Args:
+        text_embeds(`jnp.ndarray` of shape `(batch_size, output_dim`):
```
Comment (sanchit-gandhi) [inline src/transformers/models/clip/modeling_flax_clip.py:164]: (nit)
```suggestion
        text_embeds (`jnp.ndarray` of shape `(batch_size, output_dim`):
```

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 3]
```diff
@@ -1007,6 +1037,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
 )
 
 
+class FlaxCLIPTextModuleWithProjection(nn.Module):
+    config: CLIPTextConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.text_model = FlaxCLIPTextTransformer(self.config, dtype=self.dtype)
+        self.text_projection = nn.Dense(self.config.projection_dim, use_bias=False)
```
Comment (sanchit-gandhi) [inline src/transformers/models/clip/modeling_flax_clip.py:1046]: Projection layer should respect the `dtype` of the module
```suggestion
        self.text_projection = nn.Dense(self.config.projection_dim, use_bias=False, dtype=self.dtype)
```

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 4]
```diff
@@ -1007,6 +1037,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
...
+        return_dict: bool = True,
+    ):
+        text_outputs = self.text_model(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            deterministic=deterministic,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=True,
```
Comment (sanchit-gandhi) [inline src/transformers/models/clip/modeling_flax_clip.py:1065]: (nit) we usually set `return_dict=return_dict` (as per the arguments to the module) and then handle the outputs by indexing. See the PyTorch equivalent for an example: https://github.com/huggingface/transformers/blob/5ee9693a1c77c617ebc43ef20194b6d3b674318e/src/transformers/models/clip/modeling_clip.py#L1254

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 5]
```diff
@@ -1007,6 +1037,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
...
+
+FLAX_CLIP_TEXT_MODEL_WITH_PROJECTION_DOCSTRING = """
+    Returns:
+
+    Example:
+
+    ```python
+    >>> from transformers import AutoTokenizer, FlaxCLIPTextModelWithProjection
+
+    >>> model = FlaxCLIPTextModelWithProjection.from_pretrained("openai/clip-vit-base-patch32")
```
Comment (sanchit-gandhi) [inline src/transformers/models/clip/modeling_flax_clip.py:1094]: Nice example! Does this model already have Flax weights on the Hub? Or do we need to convert and push them?

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for adding this @pcuenca - looking clean! Good with me to add this Flax model if you require it for SD / SD XL - I'll leave it to you and Patrick to decide on whether it's an appropriate addition, but from a Flax transformers side the design is great. 

Feel free to add the model to the testing file such that it is run by the CI, just need to add the model class here: https://github.com/huggingface/transformers/blob/5ee9693a1c77c617ebc43ef20194b6d3b674318e/tests/models/clip/test_modeling_flax_clip.py#L318

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 5 above)
Comment (pcuenca) [reply src/transformers/models/clip/modeling_flax_clip.py:1094]: It does! :)

**Comment by pcuenca** [comment]
Thanks a lot for your in-depth review @sanchit-gandhi! 🙌 I couldn't get back to this PR until today, but I think I addressed all your comments and fixed an error in the example docstring. I was getting some seemingly unrelated CI failures so I just merged the latest `main` to see if they pass.

**Review by sanchit-gandhi — APPROVED** [review]
Looks good - thanks for iterating @pcuenca!

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me!

**Pull request merged**